### ISSUE TITLE
Improve keyboard navigation for forms & text elements

### ISSUE THREAD

**Issue by brainwane** [OpenTechStrategies/psm#510]
We should ensure that our forms can be navigated using the keyboard and other accessibility tools. In `pa11y` this is `Section508.N.KeyboardNav` ("When electronic forms are designed to be completed on-line, the form shall allow people using assistive technology to access the information, field elements, and functionality required for completion and submission of the form, including all directions and cues."). In HTML CodeSniffer:

> This text input element does not have a name available to an accessibility API. Valid names are: label element, title attribute, aria-label attribute, aria-labelledby attribute.
> Success Criterion:
> [4.1.2: Name, Role, Value](http://www.w3.org/TR/WCAG20/#ensure-compat-rsv)
> Suggested Techniques:
> [H91](http://www.w3.org/TR/WCAG20-TECHS/H91)

Examples where we currently do not comply:

~~~
<input id="enrollmentNumber" class="text" value="" type="text">
<form action=\"/cms/provider/enrollment/lookup\" id=\"practiceLookupForm\">\n <span>...</form>
<form action=\"\" id=\"enrollmentForm\" method=\"post\" enctype=\"multipart/form-data\">                    \n<div class...</form>
<form action=\"/cms/provider/enrollment/lookup\" id=\"practiceLookupForm\">\n                        <span>...</form>
<form id=\"paginationForm\" action=\"/cms/provider/dashboard/list/filter\" method=\"get\">\n<div class=\"filterPanel\" style...</form>
<form action=\"/cms/agent/enrollment/approve\" method=\"post\" id=\"approvalForm\">\n\n                        <div ...</form>
~~~

**Comment by brainwane** [comment]
We also have this issue across a bunch of our checkboxes, radio buttons, password inputs, select elements, etc. `pa11y` (when checking for WCAG level AA issues) notes these as `WCAG2AA.Principle4.Guideline4_1.4_1_2.H91.` level issues. Examples include:

~~~
<input class=\"checkbox\" name=\"_06_reimbursementSameAsPrimary\" type=\"checkbox\">
<input class=\"fileUpload\" size=\"10\" name=\"_13_attachment_0\" type=\"file\">
<input class=\"fileUpload\" size=\"10\" name=\"_03_attachment\" type=\"file\">
<input class=\"fileUpload\" size=\"10\" name=\"_14_copyOfHighestDegree\" type=\"file\">
<input type=\"password\" name=\"j_password\" id=\"password\">
<input onchange=\"javascript:submitFormById('enrollmentForm', '/cms/provider/enrollment/steps/rebind')\" value=\"Y\" name=\"_04_maintainsOwnPrivatePractice\" type=\"radio\">
<input onchange=\"javascript:submitFormById('enrollmentForm', '/cms/provider/enrollment/steps/rebind')\" value=\"N\" name=\"_04_maintainsOwnPrivatePractice\" checked=\"checked\" type=\"radio\">
<input onchange=\"javascript:submitFormById('enrollmentForm', '/cms/provider/enrollment/steps/rebind')\" value=\"Y\" name=\"_04_employedOrContractedByGroup\" checked=\"checked\" type=\"radio\">
<input onchange=\"javascript:submitFormById('enrollmentForm', '/cms/provider/enrollment/steps/rebind')\" value=\"N\" name=\"_04_employedOrContractedByGroup\" type=\"radio\">
<input id=\"agreementDocumentTitleFilterText\" class=\"normalInput\" type=\"text\">
<input id=\"userLastNameFilterText\" class=\"normalInput\" type=\"text\">
<input id=\"userFirstNameFilterText\" class=\"normalInput\" type=\"text\">
<input class=\"normalInput\" name=\"_13_licenseNumber_0\" maxlength=\"45\" type=\"text\">
~~~

**Comment by brainwane** [comment]
Example: the `Add Service Agent` form includes:

`... <form id="userForm" action="/cms/admin/createAgent" method="post" novalidate="novalidate"> <div id="addAg...</form>`

HTML CodeSniffer reminds us: "This form does not contain a submit button, which creates issues for those who cannot submit the form using the keyboard. Submit buttons are INPUT elements with type attribute "submit" or "image", or BUTTON elements with type "submit" or omitted/invalid."

Success Criterion:
[3.2.2: On Input](http://www.w3.org/TR/WCAG20/#consistent-behavior-unpredictable-change)
Suggested Techniques:
[H32](http://www.w3.org/TR/WCAG20-TECHS/H32)

**Comment by jasonaowen** [comment]
See also #65.

**Comment by jasonaowen** [comment]
I also discovered a bit more about this recently, as I was working on the CSRF stuff.

The enrollment step forms (eg, where you fill out your name and NPI and such) all have "Previous" and "Next" "buttons", which trigger some JavaScript that posts the form. It turns out the reason they do that is to be able to use the same form for both actions, as the JavaScript first updates the action with the next / previous page URL and then submits it. This allows the information to still be submitted while moving backwards, rather than having the "Previous" button lose all your changes.

That this is a pattern is, in particular, no excuse for forms outside the enrollment steps, as they have only one submit-equivalent button and can and should use a regular element rather than JavaScript.

Relevant code: search for `submitFormById(`

**Comment by jasonaowen** [comment]
There is a way that would allow us to use non-JavaScript native HTML elements for our forms: https://stackoverflow.com/a/16661115/25637 . This would require us to update our controllers and templates, and would be somewhat tedious (with all the forms we have), but is totally doable.

**Comment by PaulMorris** [comment]
I've started working on this and I think it makes sense to handle it as two separate issues:
  - "This form does not contain a submit button"  
  - "This input element does not have a name available"

I'll open separate issues for each of these.
